### DISCUSSION TITLE
How do I run a custom distutils Command without a setup.py?

### DISCUSSION THREAD

**Discussion by thrau** [pypa/setuptools#4223] [Q&A] [answered]
## Question

Is it still possible to run individual custom distutils commands without a `setup.py`?

## Background

My project https://github.com/localstack/plux provides a way to automatically generate entrypoints from code. Regenerating entrypoints has been very convenient by running `python setup.py plugins egg_info`, where `plugins` is a [custom setuptools command provided by plux](https://github.com/localstack/plux/blob/a80751d6de03ca5a0bf2a396336887e3c50a0e77/setup.cfg#L47-L52), and `egg_info` is the setuptools script to generate the `.egg-info` directory containing the entrypoints.

With the modern way of writing distributions without a `setup.py` file, and rather moving everything into `pyproject.toml` and using `build`, I no longer know how to run individual distutils/setuptools scripts. I can obviously always run `pip install -e .` again, but i'd really rather provide a command that specifically re-builds the `entry_points.txt` using the setuptools plumbing I already have.

**Comment by abravalheri** [comment]
Hi @thrau,

If you want to use commands distributed by `setuptools` and or `distutils` outside of the build process, you can still use the programmatic API for that (i.e., creating the `Distribution` and `Command` objects, setting up the correct attribute values and calling the relevant methods appropriately).

However, since the Python packaging ecosystem has started to move towards build APIs (instead of CLI calls), the scope of `setuptools` shrank, and the project no longer has as objective to:

1. Perform direct installation of packages (this is delegated to `pip`)
2. Maintain CLI capabilities for running commands.

You can read more about this shift in https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html

**Reply by thrau** [reply]
Thanks @abravalheri! That's already helpful and I can further clarify my question. I think the gist that

> The setuptools team no longer wants to be in the business of providing a command line interface

is understandable and fine! The cli is not really what I care about, I can provide my own. What I care about is the context the setuptools CLI created for me, specifically the `Distribution` object with all its properties.

I guess then my real question is: which setuptools APIs do I use, and how do I use them correctly, to get a context that looks like the one created by, say, `python setup.py egg_info`?

Or in other words:

> creating the `Distribution` and `Command` objects, setting up the correct attribute values and calling the relevant methods appropriately

How do I do that? :smile:

**Reply by abravalheri** [reply]
I think this is an example that is simple to understand:

https://github.com/pypa/setuptools/blob/v69.1.0/setuptools/tests/test_sdist.py#L169-L175

`Distribution.parse_config_files`, is also useful when your configuration is given via `pyproject.toml` or `setup.cfg`.

**Reply by thrau** [reply]
well, at least that gives me a starting point, thank you! i'll play around with this to see whether it answers my question. :-)

**Reply by abravalheri** [reply]
I believe that you can also run commands with `dist.run_command("name")`.

**Reply by thrau** [reply, marked as answer]
Based on the hints you provided, I came up with something (conceptually) like this to replace the `python setup.py plugins egg_info` command:

```python
import os
from setuptools.dist import Distribution

workdir = os.getcwd()

# prepare Distribution from workdir
config_files = ["pyproject.toml", "setup.cfg"]
config_files = [file for file in config_files if os.path.exists(file)]

if not config_files:
    raise ValueError(f"no distribution config files found in {workdir}")

dist = Distribution()
dist.parse_config_files(config_files)
if os.path.exists("setup.py"):
    # use setup.py script if available
    dist.script_name = os.path.join(workdir, "setup.py")
else:
    # else use a bogus file (seems to work regardless)
    dist.script_name = config_files[0]

# run commands
dist.run_command("plugins")
dist.run_command("egg_info")
```
